snuba/querylog/query_metadata.py

```diff
@@ -15,6 +22,107 @@ class QueryStatus(Enum):
     TIMEOUT = "timeout"
 
 
+QUERY_ERROR_MAPPINGS = {
+    ErrorCodes.TOO_SLOW: QueryStatus.TIMEOUT,
+    ErrorCodes.TIMEOUT_EXCEEDED: QueryStatus.TIMEOUT,
+    ErrorCodes.SOCKET_TIMEOUT: QueryStatus.TIMEOUT,
+    ErrorCodes.NETWORK_ERROR: QueryStatus.TIMEOUT,
+}
+
+
+def get_query_status_from_error_codes(code: ErrorCodes) -> QueryStatus | None:
+    return QUERY_ERROR_MAPPINGS.get(code)
+
+
+class SLOStatus(Enum):
+    """
+    The different statuses we return for a request.
+
+    TODO: This will replace QueryStatus, but both exist as we cut over.
+    """
+
+    # Successfully returned a response
+    SUCCESS = "success"
+    # Failed validation
+    INVALID_REQUEST = "invalid-request"
+    # A type mismatch between arguments, ILLEGAL_TYPE_OF_ARGUMENT/TYPE_MISMATCH
+    INVALID_TYPING = "invalid-typing"
+    # Rejected because of one of the product based rate limiters
+    RATE_LIMITED = "rate-limited"
+    # Rejected by a table rate limiter
+    TABLE_RATE_LIMITED = "table-rate-limited"
+    # If the thread responsible for executing a query and writing it to the cache times out
+    CACHE_SET_TIMEOUT = "cache-set-timeout"
+    # The thread waiting for a cache result to be written times out
+    CACHE_WAIT_TIMEOUT = "cache-wait-timeout"
+    # Clickhouse predicted the query would time out. TOO_SLOW
+    PREDICTED_TIMEOUT = "predicted-timeout"
+    # Clickhouse timeout, TIMEOUT_EXCEEDED
+    CLICKHOUSE_TIMEOUT = "clickhouse-timeout"
+    # Network isssues, SOCKET_TIMEOUT NETWORK_ERROR
+    NETWORK_TIMEOUT = "network-timeout"
+    # Query used too much memory, MEMORY_LIMIT_EXCEEDED
+    MEMORY_EXCEEDED = "memory-exceeded"
+    # Any other error
+    ERROR = "error"
+
+
+class SLO(Enum):
+    FOR = "for"  # These are not counted against our SLO
+    AGAINST = "against"  # These are counted against our SLO
```

Comment:
Should we model these the way HTTP error codes are laid out? Error codes from ranges 400-499 are client side errors. Ranges 500-599 are server side errors.

In a similar vein, should we assign fixed range numbers for all product concerns and another fixed range numbers for all snuba concerns?

I believe that would make the overall system simpler and categorizing new errors in the future easier as well.

Reply:
I don't know how I feel about this idea. It does make it easier to calculate whether a status is for/against the SLO, but I'm not sure it makes it easier to maintain. If a status changes how it impacts the SLO, does that mean we change its number? Does that mean our data is inconsistent?

While this would get rid of some of the mappings, it would also mean keeping a mapping of every error to an arbitrary number.

Reply:
`SUCCESS` / `FAIL` instead of FOR and AGAINST as they clearly mean which one is good and which one is bad ?

Reply:
I think I got caught up in the semantics. A request that doesn't count against our SLO might actually be more bad than one that does. For example, a completely invalid query vs. a table rate limited query.